src/atomWithQuery.ts

```diff
@@ -0,0 +1,156 @@
+): Atom<Promise<QueryObserverResult<TData, TError>>>
+export function atomWithQuery<
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+  TSuspense extends boolean = false,
+>(
+  getOptions: (
+    get: Getter
+  ) => QueryObserverOptions<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  > & { suspense?: TSuspense },
+  getQueryClient: (get: Getter) => QueryClient = (get) => get(queryClientAtom)
+) {
+  const observerCacheAtom = atom(
+    () =>
+      new WeakMap<
+        QueryClient,
+        QueryObserver<TQueryFnData, TError, TData, TQueryData, TQueryKey>
+      >()
+  )
+
+  const observerAtom = atom((get) => {
+    const options = getOptions(get)
+    const client = getQueryClient(get)
+    const observerCache = get(observerCacheAtom)
+    const observer = observerCache.get(client)
+    if (observer) {
+      ;(observer as any)[IN_RENDER] = true
+      observer.setOptions(options)
+      delete (observer as any)[IN_RENDER]
+      return observer
+    }
+    const defaultedOptions = client.defaultQueryOptions(options)
+    const newObserver = new QueryObserver(client, defaultedOptions)
+    observerCache.set(client, newObserver)
+    return newObserver
+  })
+  const IN_RENDER = Symbol()
+  const dataAtom = atom((get) => {
+    const options = getOptions(get)
+    const observer = get(observerAtom)
+
+    const source = make<QueryObserverResult<TData, TError>>(({ next }) => {
+      const callback = (result: QueryObserverResult<TData, TError>) => {
+        const notifyResult = () => next(result)
+        if ((observer as any)[IN_RENDER]) {
+          Promise.resolve().then(notifyResult)
+        } else {
+          notifyResult()
+        }
+      }
+
+      const unsubscribe = observer.subscribe(callback)
+      callback(observer.getCurrentResult())
+      return () => unsubscribe()
+    })
+
+    const observable = pipe(
+      source,
+      // tap((state) => console.log(JSON.stringify({ state }))),
+      filter((state) => {
+        if (options.suspense) {
+          if (
+            (state.isSuccess && state.data !== undefined) ||
+            (state.isError && !isCancelledError(state.error))
+          ) {
+            return true
+          } else {
+            return false
+          }
+        }
+        return true
+      }),
+      toObservable
+    )
+
+    const observableAtom = atomWithObservable(() => observable)
+
+    return observableAtom
+  })
+
+  const returnResultData = (result: any) => {
+    if (result.error) {
+      throw result.error
+    }
+    return result
+  }
+
+  const resultAtom = atom((get) => {
+    const finalAtom = get(dataAtom)
+    const result = get(finalAtom)
+
+    if (result instanceof Promise) {
+      return result.then(returnResultData)
+    } else {
+      return returnResultData(result)
+    }
+  })
+
+  return resultAtom
+}
```

Comment:
Is the implementation mostly similar to `atomsWithQuery`? Any fundamental difference?

Reply:
Implementation would be very close to atomsWithQuery with some differences in how we handle features like suspense, keepPreviousData (I've been working on this locally not in the PR yet). I'll update this PR soon.

I didn't have an opinion on how to handle those but having tested things out quite a bit more in the past week, I think we should keep things consistent with what users expect from tanstack. Refetch shouldn't trigger suspense. Suspense should be opt-in, probably with atomWithSuspenseQuery (for v5) like you mentioned. Let me know what you think.

Reply:
> I think we should keep things consistent with what users expect from tanstack. 

I agree with it. That's the major goal of the overhaul #42.